client/my-sites/stats/activity-log/index.jsx

```diff
@@ -118,6 +119,18 @@ class ActivityLog extends Component {
 		if ( ! prevProps.rewindState.rewind && this.props.rewindState.rewind ) {
 			this.findExistingRewind( this.props );
 		}
+
+		if ( ! isEqual( prevProps.filter, this.props.filter ) ) {
+			const query = filterStateToQuery( this.props.filter );
+			const queryString = query
+				.map( pair => pair.join( '=' ) )
+				.join( '&' )
+				.replace( /page=1&/, '' );
+
+			queryString === 'page=1'
+				? page( window.location.pathname + window.location.hash )
+				: page( window.location.pathname + '?' + queryString + window.location.hash );
+		}
```

Comment:
@jsnajdr @samouri @gwwar @ockham I am very conflicted on where to put this code. I want the query args to update as a result of changing filter state but I don't want to mangle the URL if the change happens as some kind of delayed callback.

originally I put it in the navigation middleware since it affects the URL and maybe I feel better putting it there because then we don't have to smush concerns here in this component, but it's intrinsically related to this component which is why I pulled it back.

any thoughts? opinions?

in this code we're just transforming internal state into a URL that can be used to come back to the same state in the UI. the funny stuff with `page=1` is just a nicety since we don't need to indicate the first page - it's assumed if no other pages are selected

Reply:
I don't know what's the right solution in Calypso right now, but we can take some guidance from React Router: there, each component that acts as a route handler receives a few props, two of them are `location` and `history`.

`location` is the parsed URL and can be used to avoid accessing `window.location`. Just look at `this.props.location.search` if you want to read the query params.

`history` can be used to navigate away. Your code could be implemented as:
```js
this.props.history.push( { search: queryString === 'page=1' ? '' : '?' + queryString } );
```
You can update just one part of the URL (`search` in this case) without worrying about other parts, which is very convenient.

I use this in ICFY to update the branch name in query string as branches get selected from a select box:
https://github.com/jsnajdr/icfy/blob/master/client/src/BranchView.js#L111

Maybe we can implement something similar with `page` and even have Redux actions for that? I think some Redux <=> page bridge already exists.

Reply:
the navigation middleware is a start to some bridge. this is like what, top priority # 14825? yeah I'd love to see us do more with that.

the HOC is an interesting idea. how do you feel about using that vs. middleware for modifying the URL?

my gosh I could even see some kind of HOC with a supplied function to determine the location bar, but that could also get quite verbose. the nice thing about the middleware is that there's very little code and all routing code is in one place. the obvious downside is that if not done carefully we could mangle the URL while on another page

Reply:
You mean a HoC like `withRouting( Component )` that supplies the `location` and `history` props to the wrapped component? That certainly looks like a good idea.

There's also `lib/url-search` HoC that modifies the URL query string as the wrapped component does search. It directly calls `page( ... )` when search query is updated.

Reply:
interesting. I might play around with a custom HOC for just this specific purpose, of filtering the ActivityLog.

Reply:
I think it's a good solution for now 👍Note that we can evolve our routing/navigation handling over time, if we find this insufficient.